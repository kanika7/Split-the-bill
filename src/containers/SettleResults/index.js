import React, { Component } from "react";
import PropTypes from "prop-types";
import { OWES, AVG_EXPENSE } from "../../constants";

import "./styles/index.scss";

class SettleResults extends Component {
  static debtArr = [];
  constructor(props) {
    super(props);
    this.state = {
      expenseArr: [],
      avgAmount: ""
    };
  }

  componentDidMount() {
    this.settleUp(this.props.expenses);
  }

  // Calculating average amount for each user.
  calculateAverage = entries => {
    const nameAndAmountSpent = {};
    const { totalAmount } = this.props;
    let totalUsers = 0;
    entries.forEach(entry => {
      if (!nameAndAmountSpent[entry.name]) {
        nameAndAmountSpent[entry.name] = entry.amount;
        totalUsers++;
      } else {
        nameAndAmountSpent[entry.name] += entry.amount;
      }
    });
    const perHeadAmount = totalAmount / totalUsers;
    return { nameAndAmountSpent, perHeadAmount };
  };

  // called for executing split of bills
  settleUp = entries => {
    const { perHeadAmount, nameAndAmountSpent } = this.calculateAverage(entries);
    const result = [];
    let str = "";

    for (let name in nameAndAmountSpent) {
      nameAndAmountSpent[name] -= perHeadAmount;
    }
    for (let i in nameAndAmountSpent) {
      if (nameAndAmountSpent[i] === 0) continue;
      // will only run for users who have to get money back
      if (nameAndAmountSpent[i] > 0) {
        for (let j in nameAndAmountSpent) {
          //will run for users who have to give the money to others
          if (nameAndAmountSpent[j] < 0) {
            if (nameAndAmountSpent[i] === -nameAndAmountSpent[j]) {
              str = `${j} ${OWES} ${i} ${nameAndAmountSpent[i].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[i] = 0;
              nameAndAmountSpent[j] = 0;
            } else if (nameAndAmountSpent[i] > -nameAndAmountSpent[j]) {
              str = `${j} ${OWES} ${i} ${-nameAndAmountSpent[j].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[i] -= nameAndAmountSpent[j];
              nameAndAmountSpent[j] = 0;
            } else {
              str = `${j} ${OWES} ${i} ${nameAndAmountSpent[i].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[j] += nameAndAmountSpent[i];
              nameAndAmountSpent[i] = 0;
            }
          }
          if (nameAndAmountSpent[i] === 0) break;
        }
      }
    }

    this.setState({
      expenseArr: [...result],
      avgAmount: perHeadAmount
    });
  };

  render() {
    const { expenseArr, avgAmount } = this.state;

    return (
      <ul className="results-container">
        {avgAmount && (
          <li className="list-item">
            <span className="col-xs-12 col-lg-4">{`${AVG_EXPENSE} ${avgAmount.toFixed(2)}`}</span>
          </li>
        )}
        {expenseArr.map((exp, index) => (
          <li key={index} className="list-item">
            <span className="col-xs-12 col-lg-4">{exp}</span>
          </li>
        ))}
      </ul>
    );
  }
}

SettleResults.propTypes = {
  expenses: PropTypes.instanceOf(Array).isRequired,
  totalAmount: PropTypes.number.isRequired
};

export default SettleResults;
