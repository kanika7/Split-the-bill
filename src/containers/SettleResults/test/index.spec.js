import React from "react";
import { shallow } from "enzyme";
import SettleResults from "../index";

describe("<SettleResults /> Test Suit", () => {
  let wrapper;
  let instance;
  const props = {
    expenses: [
      {
        username: "test abc",
        amount: "20",
        description: "test dec"
      }
    ],
    totalAmount: "200"
  };

  function renderShallow() {
    wrapper = shallow(<SettleResults {...props} />);
  }

  it("renders snapshot", () => {
    renderShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("handles calculateAverage to find the total and per head expense", () => {
    const props = {
      expenses: [
        {
          name: "jane",
          amount: 20,
          description: "test dec"
        },
        {
          name: "john",
          amount: 10,
          description: "test dec"
        },
        {
          name: "john",
          amount: 20,
          description: "test dec"
        }
      ],
      totalAmount: "50"
    };
    const expectedObj = {
      nameAndAmountSpent: {
        jane: 20,
        john: 30
      },
      perHeadAmount: 25
    };
    wrapper = shallow(<SettleResults {...props} />);
    instance = wrapper.instance();
    const spy = jest.spyOn(instance, "calculateAverage");
    instance.calculateAverage([...props.expenses]);
    expect(spy).toHaveReturnedWith(expectedObj);
  });

  it("handles settleUp to handle a user with 0 expense", () => {
    const props = {
      expenses: [
        {
          name: "jane",
          amount: 20,
          description: "test dec"
        },
        {
          name: "dave",
          amount: 0,
          description: "test dec"
        },
        {
          name: "john",
          amount: 70,
          description: "test dec"
        }
      ],
      totalAmount: "90"
    };
    wrapper = shallow(<SettleResults {...props} />);
    instance = wrapper.instance();
    jest.spyOn(instance, "settleUp");
    instance.settleUp([...props.expenses]);
    expect(wrapper.state()).toEqual({
      expenseArr: ["jane owes john 10", "dave owes john 30"],
      avgAmount: 30
    });
  });

  it("handles settleUp to handle a user where his expense is already paid as per Average amount to be shared", () => {
    const props = {
      expenses: [
        {
          name: "jane",
          amount: 30,
          description: "test dec"
        },
        {
          name: "dave",
          amount: 5,
          description: "test dec"
        },
        {
          name: "john",
          amount: 55,
          description: "test dec"
        }
      ],
      totalAmount: "90"
    };
    wrapper = shallow(<SettleResults {...props} />);
    instance = wrapper.instance();
    jest.spyOn(instance, "settleUp");
    instance.settleUp([...props.expenses]);
    expect(wrapper.state()).toEqual({
      expenseArr: ["dave owes john 25.00"],
      avgAmount: 30
    });
  });

  it("handles settleUp condition where the person A has money which can fullfill one debter's complete owned amount and A is still left with some more amout to give", () => {
    const props = {
      expenses: [
        {
          name: "jane",
          amount: 100,
          description: "test dec"
        },
        {
          name: "dave",
          amount: 100,
          description: "test dec"
        },
        {
          name: "john",
          amount: 50,
          description: "test dec"
        }
      ],
      totalAmount: "250"
    };
    wrapper = shallow(<SettleResults {...props} />);
    instance = wrapper.instance();
    jest.spyOn(instance, "settleUp");
    instance.settleUp([...props.expenses]);
    expect(wrapper.state()).toEqual({
      expenseArr: ["john owes jane 16.67", "john owes dave 16.67"],
      avgAmount: 83.33333333333333
    });
  });
});
