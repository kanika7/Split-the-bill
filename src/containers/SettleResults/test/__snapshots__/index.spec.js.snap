// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettleResults /> Test Suit renders snapshot 1`] = `
<ul
  className="results-container"
>
  <li
    className="list-item"
  >
    <span
      className="col-xs-12 col-lg-4"
    >
      Per Head Expense is:  200.00
    </span>
  </li>
</ul>
`;
